Flow.Launcher.Infrastructure/Image/ThumbnailReader.cs

```diff
@@ -115,39 +44,56 @@ public static BitmapSource GetThumbnail(string fileName, int width, int height,
             finally
             {
                 // delete HBitmap to avoid memory leaks
-                DeleteObject(hBitmap);
+                PInvoke.DeleteObject(hBitmap);
             }
         }
-        
-        private static IntPtr GetHBitmap(string fileName, int width, int height, ThumbnailOptions options)
+
+        private static unsafe HBITMAP GetHBitmap(string fileName, int width, int height, ThumbnailOptions options)
         {
-            IShellItem nativeShellItem;
-            Guid shellItem2Guid = new Guid(IShellItem2Guid);
-            int retCode = SHCreateItemFromParsingName(fileName, IntPtr.Zero, ref shellItem2Guid, out nativeShellItem);
+            var retCode = PInvoke.SHCreateItemFromParsingName(
+                fileName,
+                null,
+                GUID_IShellItem,
+                out var nativeShellItem);
 
-            if (retCode != 0)
+            if (retCode != HRESULT.S_OK)
                 throw Marshal.GetExceptionForHR(retCode);
 
-            NativeSize nativeSize = new NativeSize
+            if (nativeShellItem is not IShellItemImageFactory imageFactory)
             {
-                Width = width,
-                Height = height
-            };
+                Marshal.ReleaseComObject(nativeShellItem);
+                nativeShellItem = null;
+                throw new InvalidOperationException("Failed to get IShellItemImageFactory");
+            }
 
-            IntPtr hBitmap;
-            HResult hr = ((IShellItemImageFactory)nativeShellItem).GetImage(nativeSize, options, out hBitmap);
+            SIZE size = new SIZE
+            {
+                cx = width,
+                cy = height
+            };
 
-            // if extracting image thumbnail and failed, extract shell icon
-            if (options == ThumbnailOptions.ThumbnailOnly && hr == HResult.ExtractionFailed)
+            HBITMAP hBitmap = default;
+            try
             {
-                hr = ((IShellItemImageFactory) nativeShellItem).GetImage(nativeSize, ThumbnailOptions.IconOnly, out hBitmap);
+                try
+                {
+                    imageFactory.GetImage(size, (SIIGBF)options, &hBitmap);
```

Comment:
do cswin32 provides the automatic generation for GetImage?

Reply:
Any problem for current codes? Or any difference it can make? (Sorry, I am not quite familiar with this...)

Reply:
no just curious

Reply:
I just follow the original style. And I have tested it can work. So I have not thought about this.